sql/core/src/test/scala/org/apache/spark/sql/InSetBenchmark.scala

```diff
@@ -0,0 +1,213 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.benchmark.Benchmark
+import org.apache.spark.sql.catalyst.expressions.In
+import org.apache.spark.sql.catalyst.expressions.InSet
+import org.apache.spark.sql.execution.benchmark.SqlBasedBenchmark
+import org.apache.spark.sql.functions.{array, struct}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+
+/**
+ * A benchmark that compares the performance of [[In]] and [[InSet]] expressions.
+ *
+ * To run this benchmark:
+ * {{{
+ *   1. without sbt: bin/spark-submit --class <this class> <spark sql test jar>
+ *   2. build/sbt "sql/test:runMain <this class>"
+ *   3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
+ *      Results will be written to "benchmarks/InSetBenchmark-results.txt".
+ * }}}
+ */
+object InSetBenchmark extends SqlBasedBenchmark {
+
+  import spark.implicits._
+
+  def byteBenchmark(numItems: Int, numRows: Long, minNumIters: Int): Benchmark = {
+    val name = s"$numItems bytes"
+    val values = (1 to numItems).map(v => s"CAST($v AS tinyint)")
+    val df = spark.range(1, numRows).select($"id".cast(ByteType))
+    benchmark(name, df, values, numRows, minNumIters)
+  }
+
+  def shortBenchmark(numItems: Int, numRows: Long, minNumIters: Int): Benchmark = {
+    val name = s"$numItems shorts"
+    val values = (1 to numItems).map(v => s"CAST($v AS smallint)")
+    val df = spark.range(1, numRows).select($"id".cast(ShortType))
+    benchmark(name, df, values, numRows, minNumIters)
+  }
+
+  def intBenchmark(numItems: Int, numRows: Long, minNumIters: Int): Benchmark = {
+    val name = s"$numItems ints"
+    val values = 1 to numItems
+    val df = spark.range(1, numRows).select($"id".cast(IntegerType))
+    benchmark(name, df, values, numRows, minNumIters)
```

Comment:
nit. Shall we invoke `run` here intead of `intBenchmark(...).run()`? It seems to be repeated multiple times.
```
- benchmark(name, df, values, numRows, minNumIters)
+ benchmark(name, df, values, numRows, minNumIters).run()
```

Reply:
We can do this. Shall we rename `intBenchmark` to `runIntBenchmark` then? There is no consistency in existing benchmarks, unfortunately.